src/Raven.Server/SqlMigration/RavenWriter.cs

```diff
@@ -0,0 +1,165 @@
+                            throw;
+                        }
+                    }
+                }
+                
+                OnTableWritten?.Invoke(table.Name, (double) sw.ElapsedMilliseconds / 1000);
+            }
+            await EnqueueCommands();
+        }
+
+        private const int BatchSize = 1000; // TODO:  what's the most efficient batch size?
+        private readonly List<BatchRequestParser.CommandData> _documentCommands = new List<BatchRequestParser.CommandData>();
+        private readonly List<AttachmentHandler.MergedPutAttachmentCommand> _attachmentCommands = new List<AttachmentHandler.MergedPutAttachmentCommand>();
+        private readonly List<IDisposable> _toDispose = new List<IDisposable>();
+
+        private async Task InsertDocument(RavenDocument ravenDocument, Dictionary<string, byte[]> attachments)
+        {
+            BlittableJsonReaderObject document;
+
+            try
+            {
+                document = _context.ReadObject(ravenDocument, ravenDocument.Id, BlittableJsonDocumentBuilder.UsageMode.ToDisk);
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException($"Cannot build raven document from table '{ravenDocument.TableName}'. Raven document id is: {ravenDocument.Id}", e);
+            }
+
+            _documentCommands.Add(new BatchRequestParser.CommandData
+            {
+                Type = CommandType.PUT,
+                Document = document,
+                Id = ravenDocument.Id
+            });
+
+            foreach (var attachment in attachments)
+            {
+                var streamsTempFile = _database.DocumentDatabase.DocumentsStorage.AttachmentsStorage.GetTempFile("put");
+                var stream = streamsTempFile.StartNewStream();
+                
+                var ms = new MemoryStream(attachment.Value);
+                var hash = await AttachmentsStorageHelper.CopyStreamToFileAndCalculateHash(_context, ms, stream, _database.DocumentDatabase.DatabaseShutdown);
+
+                _attachmentCommands.Add(new AttachmentHandler.MergedPutAttachmentCommand
+                {
+                    Database = _database.DocumentDatabase,
+                    DocumentId = ravenDocument.Id,
+                    Name = attachment.Key,
+                    Stream = stream,
+                    Hash = hash,
+                    ContentType = ""
+                });
+                _toDispose.Add(stream);
+                _toDispose.Add(streamsTempFile);
+             }
+
+            if (_documentCommands.Count % BatchSize == 0)
+            {
+                await EnqueueCommands();
+                _documentCommands.Clear();
+                _attachmentCommands.Clear();
+            }
+        }
+
+        private async Task EnqueueCommands()
+        {
+            using (var command = new BatchHandler.MergedBatchCommand { Database = _database.DocumentDatabase })
```

Comment:
You are going to wait for each independently? 
Enqueue them all and wait for all of them in one shot.